daft/runners/ray_runner.py

```diff
@@ -475,56 +477,79 @@ def build_partitions(
     return [metadatas, *partitions]
 
 
+@dataclasses.dataclass(frozen=True)
+class PartitionTaskContext:
+    job_id: str
+    task_id: str
+    stage_id: int
+
+
 # Give the same function different names to aid in profiling data distribution.
 
 
+@ray_tracing.ray_remote_traced
 @ray.remote
 def single_partition_pipeline(
+    task_context: PartitionTaskContext,
     daft_execution_config: PyDaftExecutionConfig,
     instruction_stack: list[Instruction],
     partial_metadatas: list[PartitionMetadata],
     *inputs: MicroPartition,
 ) -> list[list[PartitionMetadata] | MicroPartition]:
     with execution_config_ctx(
         config=daft_execution_config,
+    ), ray_tracing.collect_ray_task_metrics(
+        task_context.job_id, task_context.task_id, task_context.stage_id, daft_execution_config
     ):
         return build_partitions(instruction_stack, partial_metadatas, *inputs)
 
 
+@ray_tracing.ray_remote_traced
 @ray.remote
 def fanout_pipeline(
+    task_context: PartitionTaskContext,
     daft_execution_config: PyDaftExecutionConfig,
     instruction_stack: list[Instruction],
     partial_metadatas: list[PartitionMetadata],
     *inputs: MicroPartition,
 ) -> list[list[PartitionMetadata] | MicroPartition]:
-    with execution_config_ctx(config=daft_execution_config):
+    with execution_config_ctx(config=daft_execution_config), ray_tracing.collect_ray_task_metrics(
+        task_context.job_id, task_context.task_id, task_context.stage_id, daft_execution_config
+    ):
         return build_partitions(instruction_stack, partial_metadatas, *inputs)
 
 
+@ray_tracing.ray_remote_traced
 @ray.remote(scheduling_strategy="SPREAD")
 def reduce_pipeline(
+    task_context: PartitionTaskContext,
     daft_execution_config: PyDaftExecutionConfig,
     instruction_stack: list[Instruction],
     partial_metadatas: list[PartitionMetadata],
     inputs: list,
 ) -> list[list[PartitionMetadata] | MicroPartition]:
     import ray
 
-    with execution_config_ctx(config=daft_execution_config):
+    with execution_config_ctx(config=daft_execution_config), ray_tracing.collect_ray_task_metrics(
+        task_context.job_id, task_context.task_id, task_context.stage_id, daft_execution_config
```

Comment:
you can just pass in the `task_context` instead of desugering

Reply:
There's a circular dependency here unless I move `PartitionTaskContext` into a separate module (since `ray_runner.py` depends on `ray_metrics.py` and `ray_tracing.py`, but `PartitionTaskContext` is defined in `ray_runner.py`)

I decided to keep the `PartitionTaskContext` as a runtime-only concept to simplify things and avoid creating yet another module